### PR TITLE
Rename Infograph to InfoGraphStar, add documentation and fixes

### PR DESCRIPTION
**Pull request by tonydavis629**
# Pull Request Template

## Description

This PR renames InfoGraph to InfoGraphStar, which is used for supervised and semi-supervised training while InfoGraph is only used for unsupervised training.

## Type of change

Please check the option that is related to your PR.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [x] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
- [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
- [x] Run `mypy -p deepchem` and check no errors
- [x] Run `flake8 <modified file> --count` and check no errors
- [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/torch_models/infograph.py [hunk 1]
```diff
@@ -101,81 +291,138 @@ class InfoGraphModel(ModularTorchModel):
...
+        """
+        Builds the components of the InfoGraphStar model. InfoGraphStar works by maximizing the mutual information between the graph-level representation and the representations of substructures of different scales.
+
+        It does this by producing graph-level encodings and substructure encodings, and then using a discriminator to classify if they are from the same molecule or not.
+
+        The encoder is a graph convolutional network that produces the graph-level encodings and substructure encodings.
+
+        In a supervised training mode, only 1 encoder is used and the encodings are not compared, while in a semi-supvervised training mode they are different in order to prevent negative transfer from the pretraining stage.
+
+        The local discriminator is a multilayer perceptron that classifies if the substructure encodings are from the same molecule or not while the global discriminator classifies if the graph-level encodings are from the same molecule or not.
```
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:340]: My apologies for the confusion, but I meant that the explicit names of the components should be documented in the API. The description here is useful too though! We should have both

File: deepchem/models/torch_models/infograph.py [hunk 2]
```diff
@@ -101,81 +291,138 @@ class InfoGraphModel(ModularTorchModel):
     >>> ds = NumpyDataset(X, y, w)
     >>> num_feat = max([ds.X[i].num_node_features for i in range(len(ds))])
     >>> edge_dim = max([ds.X[i].num_edge_features for i in range(len(ds))])
-    >>> model = InfoGraphModel(num_feat, edge_dim, 15, use_unsup_loss=True, separate_encoder=True)
+    >>> model = InfoGraphStarModel(num_feat, edge_dim, 15, training_mode='semisupervised')
     >>> loss = model.fit(ds, nb_epoch=1)
```
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:295]: In the original infograph* paper, they provide the ability to learn from different sets of labeled and un-labeled examples. Do we have that capability in our implementation? We should add in that capability if we don't have it, since otherwise we can't use a larger unlabeled set to bootstrap training on a small labeled dataset

**Review by rbharath — COMMENTED** [review]
Re-reading the original paper, it may be that we're missing a bit of important functionality (the ability to have different sets of labeled/unlabeled data for semi-supervised learning). Also possible this is my misunderstanding though! Let's discuss offline since that might be easier than a long github thread

File: deepchem/models/torch_models/infograph.py [hunk 3]
```diff
@@ -101,81 +291,138 @@ class InfoGraphModel(ModularTorchModel):
...
+        self.prior = False
+        self.gamma = .1
         self.num_features = num_features
-        self.use_unsup_loss = use_unsup_loss
         self.measure = measure
         self.average_loss = average_loss
+        self.training_mode = training_mode
+        if training_mode == 'supervised':
+            self.embedding_dim = embedding_dim
+            self.use_unsup_loss = False
```
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:318]: We should remove `self.use_unsup_loss`. It's no longer an argument and is a duplicate of information in `training_mode`

Comment (rbharath) [reply deepchem/models/torch_models/infograph.py:318]: We should remove `self.separate_encoder` for the same reason

File: deepchem/models/torch_models/infograph.py [hunk 4]
```diff
@@ -217,16 +445,16 @@ def unsup_sup_loss(self, inputs, labels, weights):
         loss = self._global_global_loss(g_enc, g_enc1)
         return loss
 
-    def _local_global_loss(self, l_enc, g_enc, batch):
+    def _local_global_loss(self, l_enc, g_enc, index):
```
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:448]: This should be factored out to be a loss function (we can remove the self.device, self.measure, self.average_loss calls by making appropriate flags). This is a general mathematical loss term that might be useful in different contexts

File: deepchem/models/torch_models/infograph.py [hunk 5]
```diff
@@ -217,16 +445,16 @@ def unsup_sup_loss(self, inputs, labels, weights):
         loss = self._global_global_loss(g_enc, g_enc1)
```
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:445]: We should factor out `_global_global_loss` to be a loss function as well.

File: deepchem/models/losses.py [hunk 6]
```diff
@@ -465,6 +470,216 @@ def loss(inputs):
...
 
 
+class MolGraphConvMILoss():
+    """
+    Losses used to maximize the mutual information between molecular graph encodings.
+    Encodings can be generated from molecular graphs using a graph convolutional network, and the mutual information is calculated between subgraphs (local encodings) and the entire molecule (global encodings) in order to enable unsupervised training.
+
+    This class supports only PyTorch.
+
+    Parameters:
```
Comment (rbharath) [inline deepchem/models/losses.py:480]: Formatting here is off:

```
Parameters
----------
```

File: deepchem/models/losses.py [hunk 7]
```diff
@@ -465,6 +470,216 @@ def loss(inputs):
         return loss
 
 
+class MolGraphConvMILoss():
+    """
+    Losses used to maximize the mutual information between molecular graph encodings.
+    Encodings can be generated from molecular graphs using a graph convolutional network, and the mutual information is calculated between subgraphs (local encodings) and the entire molecule (global encodings) in order to enable unsupervised training.
```
Comment (rbharath) [inline deepchem/models/losses.py:476]: Can you add a usage example?

File: deepchem/models/losses.py [hunk 8]
```diff
@@ -465,6 +470,216 @@ def loss(inputs):
         return loss
 
 
+class MolGraphConvMILoss():
```
Comment (rbharath) [inline deepchem/models/losses.py:473]: Can you add this loss to the docs?

File: deepchem/models/losses.py [hunk 9]
```diff
@@ -465,6 +470,216 @@ def loss(inputs):
...
+    average_loss: bool
+        Whether to average the loss over the batch
+    """
+
+    def __init__(self, device, measure='JSD', average_loss=True):
+        self.device = device
+        self.measure = measure
+        self.average_loss = average_loss
+
+    def local_global_loss(self, l_enc, g_enc, index):
```
Comment (rbharath) [inline deepchem/models/losses.py:495]: This class isn't following the convention the other losses follow. This function should be called `_create_pytorch_loss`. See https://github.com/deepchem/deepchem/blob/master/deepchem/models/losses.py#L208 for a representative example

File: deepchem/models/losses.py [hunk 10]
```diff
@@ -465,6 +470,216 @@ def loss(inputs):
...
+
+        Parameters:
+        ----------
+        l_enc: torch.Tensor
+            Local feature map from the encoder.
+        g_enc: torch.Tensor
+            Global features from the encoder.
+        index: torch.Tensor
+            Index of the graph that each node belongs to.
+        device: torch.device
```
Comment (rbharath) [inline deepchem/models/losses.py:508]: Some of these are no longer arguments for this function and can be removed (device/measure/etc)

File: deepchem/models/losses.py [hunk 11]
```diff
@@ -465,6 +470,216 @@ def loss(inputs):
...
+        self.average_loss = average_loss
+
+    def local_global_loss(self, l_enc, g_enc, index):
+        """
+        Compares the local and global encodings of molecular graphs and returns the loss between them based on the measure specified.
+        Global in this case refers to the global encoding of the entire molecule, and local refers to the encoding of subgraphs.
+
+        Parameters:
+        ----------
+        l_enc: torch.Tensor
```
Comment (rbharath) [inline deepchem/models/losses.py:502]: What is the expected shape of `l_enc`, `g_enc`, `index`?

File: deepchem/models/losses.py [hunk 12]
```diff
@@ -465,6 +470,216 @@ def loss(inputs):
...
+        E_pos = self.get_positive_expectation(res * pos_mask, self.measure,
+                                              self.average_loss)
+        E_pos = (E_pos * pos_mask).sum() / pos_mask.sum()
+        E_neg = self.get_negative_expectation(res * neg_mask, self.measure,
+                                              self.average_loss)
+        E_neg = (E_neg * neg_mask).sum() / neg_mask.sum()
+
+        return E_neg - E_pos
+
+    def global_global_loss(self, g_enc, g_enc1):
```
Comment (rbharath) [inline deepchem/models/losses.py:541]: This should be a separate loss and not in the same loss class.

File: deepchem/models/losses.py [hunk 13]
```diff
@@ -465,6 +470,216 @@ def loss(inputs):
...
+        res = torch.mm(g_enc, g_enc1.t())
+
+        E_pos = self.get_positive_expectation(res * pos_mask)
+        E_pos = (E_pos * pos_mask).sum() / pos_mask.sum()
+        E_neg = self.get_negative_expectation(res * neg_mask)
+        E_neg = (E_neg * neg_mask).sum() / neg_mask.sum()
+
+        return E_neg - E_pos
+
+    def get_positive_expectation(self, p_samples):
```
Comment (rbharath) [inline deepchem/models/losses.py:579]: This should be a utility function and not a method (instead of `self`, just accept `measure` as a direct argument)

File: deepchem/models/losses.py [hunk 14]
```diff
@@ -465,6 +470,216 @@ def loss(inputs):
...
+            Ep = p_samples
+        else:
+            raise ValueError('Unknown measure: {}'.format(self.measure))
+
+        if self.average_loss:
+            return Ep.mean()
+        else:
+            return Ep
+
+    def get_negative_expectation(self, q_samples):
```
Comment (rbharath) [inline deepchem/models/losses.py:620]: Same for this method; should be a utility function

File: deepchem/models/torch_models/infograph.py [hunk 15]
```diff
@@ -101,93 +293,169 @@ class InfoGraphModel(ModularTorchModel):
...
+
+        The encoder is a graph convolutional network that produces the graph-level encodings and substructure encodings.
+
+        In a supervised training mode, only 1 encoder is used and the encodings are not compared, while in a semi-supvervised training mode they are different in order to prevent negative transfer from the pretraining stage.
+
+        The local discriminator is a multilayer perceptron that classifies if the substructure encodings are from the same molecule or not while the global discriminator classifies if the graph-level encodings are from the same molecule or not.
+
+        Components list and type:
+        ------------------------
+        Encoder: InfoGraphEncoder
```
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:341]: These names should match the exact strings used internally. I think we will later want to expand the modular torch model API to allow fetching the outputs of particular components

File: deepchem/models/torch_models/infograph.py [hunk 16]
```diff
@@ -101,93 +293,169 @@ class InfoGraphModel(ModularTorchModel):
...
+        if self.task == 'semisupervised':
             y, M = self.components['unsup_encoder'](inputs)
         else:
             y, M = self.components['encoder'](inputs)
         g_enc = self.components['global_d'](y)
         l_enc = self.components['local_d'](M)
 
         if self.local:
-            loss = self._local_global_loss(l_enc, g_enc, inputs.graph_index)
+            loss = self.MIloss.local_global_loss(l_enc, g_enc,
```
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:440]: We should follow existing convention for how losses behave, see https://github.com/deepchem/deepchem/blob/master/deepchem/models/losses.py#L1 and how `TorchModel` uses losses. See also my comments above on the loss class itself

**Review by rbharath — COMMENTED** [review]
Looking better! But we need one more pass. Can you break the loss out into its own small PR? We can do `InfoGraphStar` as a follow-up PR after that. This PR is getting again too large and it's hard to do a proper review.

For the loss PR, let's make sure to follow existing convention around losses. You will want to add multiple losses rather than one complicated loss function with multiple methods

File: deepchem/models/torch_models/infograph.py (hunk 2 above)
Comment (rbharath) [reply deepchem/models/torch_models/infograph.py:297]: Adding on to this previous comment, we should add a discussion in the docs about this limitation and point to the mismatch with the math in the original paper even if we don't have an easy fix for how to use two different datasets for labeled/unlabeled data

File: deepchem/models/losses.py [hunk 17]
```diff
@@ -599,11 +600,12 @@ def _create_pytorch_loss(self, measure='JSD', average_loss=True):
         import torch
 
         def loss(local_enc, global_enc, batch_graph_index):
+            device = local_enc.device
             num_graphs = global_enc.shape[0]
             num_nodes = local_enc.shape[0]
 
-            pos_mask = torch.zeros((num_nodes, num_graphs))
-            neg_mask = torch.ones((num_nodes, num_graphs))
+            pos_mask = torch.zeros((num_nodes, num_graphs)).to(device)
```
Comment (rbharath) [inline deepchem/models/losses.py:607]: I just gave an LGTM on the other PR, so we may want to rebase this one before merge

File: deepchem/models/torch_models/infograph.py (hunk 2 above)
Comment (rbharath) [reply deepchem/models/torch_models/infograph.py:300]: Quick reminder to add this discussion

File: deepchem/models/torch_models/infograph.py [hunk 18]
```diff
@@ -101,363 +296,171 @@ class InfoGraphModel(ModularTorchModel):
...
-            MultilayerPerceptron(self.embedding_dim,
-                                 self.embedding_dim, (self.embedding_dim,),
-                                 skip_connection=True),
-            'global_d':
-            MultilayerPerceptron(2 * self.embedding_dim,
-                                 self.embedding_dim, (self.embedding_dim,),
-                                 skip_connection=True),
-        }
+        """
+        Builds the components of the InfoGraphStar model. InfoGraphStar works by maximizing the mutual information between the graph-level representation and the representations of substructures of different scales.
```
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:335]: New comment here is great!

**Review by rbharath — COMMENTED** [review]
This is looking good! New implementation is clean and simple. A few minor comments but this is almost ready to go

**Review by rbharath — COMMENTED** [review]
Capturing offline discussion, decided it's OK to not mention the possibility of having a larger unsupervised set for semisupervised learning since the original paper doesn't mention this in detail and the original code release doesn't support it either.

LGTM, feel free to merge once CI is clear

**Review by rbharath — APPROVED** [review]
Hitting approve this time
